### PR TITLE
JAMES-3756 Delegation store should allow listing accounts delegated to me

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/james-project/issues/4673

### CODE REVIEW

File: server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraUsersRepositoryModule.java [hunk 1]
```diff
@@ -32,6 +32,7 @@ public interface CassandraUsersRepositoryModule {
             .withColumn(CassandraUserTable.REALNAME, DataTypes.TEXT)
             .withColumn(CassandraUserTable.PASSWORD, DataTypes.TEXT)
             .withColumn(CassandraUserTable.ALGORITHM, DataTypes.TEXT)
-            .withColumn(CassandraUserTable.AUTHORIZED_USERS, DataTypes.setOf(DataTypes.TEXT)))
+            .withColumn(CassandraUserTable.AUTHORIZED_USERS, DataTypes.setOf(DataTypes.TEXT))
+            .withColumn(CassandraUserTable.DELEGATED_USERS, DataTypes.setOf(DataTypes.TEXT)))
```
Comment (quantranhong1999) [inline server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraUsersRepositoryModule.java:36]: Maybe modify `upgrade-instruction` in this PR also so we would not forget?

File: server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraDelegationStore.java [hunk 2]
```diff
@@ -40,16 +46,31 @@ public Publisher<Username> authorizedUsers(Username baseUser) {
...
+            .then(cassandraUsersDAO.removeAllAuthorizedUsers(baseUser));
     }
 
     @Override
     public Publisher<Void> addAuthorizedUser(Username baseUser, Username userWithAccess) {
-        return cassandraUsersDAO.addAuthorizedUsers(baseUser, userWithAccess);
+        return cassandraUsersDAO.addAuthorizedUsers(baseUser, userWithAccess)
+            .then(Mono.from(cassandraUsersDAO.containsReactive(userWithAccess))
+                .filter(FunctionalUtils.identityPredicate())
+                .flatMap(authorizedUser -> cassandraUsersDAO.addDelegatedToUsers(userWithAccess, baseUser))
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraDelegationStore.java:59]: "to **achieve** atomicity".

And this is not atomicity (batch brings no such guaranty) but more eventual consistency (all actions are eventually applied, or none).

But yes, we need a (logged) batch here.

File: server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraUsersDAO.java [hunk 3]
```diff
@@ -199,6 +255,27 @@ public Flux<Username> getAuthorizedUsers(Username name) {
...
+    public Flux<Username> getDelegatedToUsers(Username name) {
+        return executor.executeSingleRow(
+                getDelegatedToUsersStatement.bind()
+                    .setString(NAME, name.asString()))
+            .mapNotNull(row -> row.getSet(DELEGATED_USERS, String.class))
+            .flatMapIterable(set -> set)
+            .map(Username::of);
+    }
+
+    // TODO, should we remove delegated user here?
```
Comment (vttranlina) [inline server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraUsersDAO.java:278]: Should we remove delegated users here?

**Review by chibenwa — APPROVED** [review]
OKish to me.

File: server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraUsersDAO.java (hunk 3 above)
Comment (quantranhong1999) [reply server/data/data-cassandra/src/main/java/org/apache/james/user/cassandra/CassandraUsersDAO.java:278]: No, not related IMO

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**